### PR TITLE
use unmerged fastq-files

### PR DESCRIPTION
**Pull request by Smeds**
merging will now be performed after alignment to better preserve information about read origin.

### Review and tests: 
- [ ] Tests pass

### CODE REVIEW

**Comment by jonca79** [comment]
In .tests you need to change config and resources

**Review by marrip — APPROVED** [review]
Looking good, just need to fix Snakefmt and what Jonas suggested.
I was also wondering a bit about `.smk`-files. I like that you included several rules in one as they are tied together but in other modules we are splitting rules into several files - what should we do? For me, it makes more sense to combine rules in one `.smk`-file in order to structure them and avoid to inflate the file count in the `rules`-dir. Another point I would like to take up is that the `merge` and `sort` rules are very specific here but they could be used in other contexts as well. Would it make more sense to make them more reusable. I encountered a couple of rules which were quite tailored where it was unnecessary and just lead to an increased amount on configuration in the main workflow which I find tedious.

**Comment by Smeds** [comment]
> Looking good, just need to fix Snakefmt and what Jonas suggested. I was also wondering a bit about `.smk`-files. I like that you included several rules in one as they are tied together but in other modules we are splitting rules into several files - what should we do? For me, it makes more sense to combine rules in one `.smk`-file in order to structure them and avoid to inflate the file count in the `rules`-dir. Another point I would like to take up is that the `merge` and `sort` rules are very specific here but they could be used in other contexts as well. Would it make more sense to make them more reusable. I encountered a couple of rules which were quite tailored where it was unnecessary and just lead to an increased amount on configuration in the main workflow which I find tedious.

snakefmt is now fixed!

dry-run will be fixed soon

I have updated the config files!

I concur that sort and merge are a bit specific. And we already have a merge smk file, but that one would have to be made more general for me to use it. 
- How do we handle logs and benchmark files if we have more general sort and merge smk files? 
-  Do we trigger them by putting all merged bam files in merge_bam or maybe by adding _unmerged_ or _unsorted_ to the filenames?
- For merging it will be a bit hard to setup input files, merging on chromosome vs merging on lane will require different functions

Any good ideas of how we solve that? @jonca79 @marrip

**Comment by marrip** [comment]
Nicely done!

About the potential misc rules:

1. For logs and benchmarks it could maybe look like:
```
rule my_rule:
    input:
        {file}.unsorted.bam,
   output:
        {file}.bam,
    log:
        {file}.bam.log,
    benchmark:
        {file}.bam.benchmark.tsv,
```
2. I agree with the tag that is needed to trigger the merging/sorting. We can view it as some kind of longer extension :wink:.
3. For the merging, it is a bit harder, I agree. We could extent the trigger extension to something like `.unmerged_chr.bam` and the appropriate function could be picked...not sure if that makes sense.

File: .tests/integration/config.yaml [hunk 1]
```diff
@@ -16,13 +16,18 @@ reference:
...
 bwa_mem:
   container: "docker://hydragenetics/bwa_mem:0.7.17"
-  extra: "-R '@RG\\tID:{sample}\\tSM:{sample}\\tPL:illumina\\tPU:{sample}' -v 1 "
 
 mark_duplicates:
   container: "docker://gmsuppsala/somatic:develop"
 
 samtools_index:
-  container: "docker://gmsuppsala/somatic:develop"
+  container: "docker://hydragenetics/common:0.0.1"
```
Comment (marrip) [inline .tests/integration/config.yaml:24]: If we use the default container, do we still need to specify that?

File: workflow/rules/common.smk [hunk 2]
```diff
@@ -14,6 +14,16 @@ from hydra_genetics.utils.misc import extract_chr
 from hydra_genetics.utils.units import *
 from hydra_genetics.utils.samples import *
 
+
+def generate_read_group(wildcards):
```
Comment (marrip) [inline workflow/rules/common.smk:18]: I would keep the functions in the bottom part of `common.smk` and have the whole config/setup in the beginning 😉

File: workflow/rules/samtools_sort.smk [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+__copyright__ = "Copyright 2021, Jonas Almlöf"
+__email__ = "patrik.smeds@scilifelab.uu.se"
+__license__ = "GPL-3"
+
+
+rule samtools_sort:
+    input:
+        "{path_file}.bam_unsorted",
+    output:
+        "{path_file}.bam",
```
Comment (marrip) [inline workflow/rules/samtools_sort.smk:14]: temp()?

File: workflow/rules/samtools_sort.smk [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    log:
+        "{path_file}.samtools_sort.log",
```
Comment (marrip) [inline workflow/rules/samtools_sort.smk:16]: It seems that we are using different systems for naming the log and benchmark files. We should agree on one way I think. Let's discuss soon.

File: workflow/schemas/resources.schema.yaml [hunk 5]
```diff
@@ -78,6 +78,25 @@ properties:
         type: string
         description: partition to use on the cluster for 'samtools_index'
 
+  samtools_sort:
```
Comment (marrip) [inline workflow/schemas/resources.schema.yaml:81]: Another suggestion I have, shouldn't we sort this level of stanzas, keeping default_resources highest up, of course?

File: .tests/integration/config.yaml (hunk 1 above)
Comment (Smeds) [reply .tests/integration/config.yaml:24]: maybe not! Let's decide on a common practice.

File: workflow/rules/common.smk (hunk 2 above)
Comment (Smeds) [reply workflow/rules/common.smk:18]: I'm up for setting guidelines for this and bottom works for me.

File: workflow/rules/samtools_sort.smk (hunk 3 above)
Comment (Smeds) [reply workflow/rules/samtools_sort.smk]: done

File: .tests/integration/config.yaml (hunk 1 above)
Comment (marrip) [reply .tests/integration/config.yaml:24]: yes please! 👍

File: workflow/schemas/resources.schema.yaml (hunk 5 above)
Comment (Smeds) [reply workflow/schemas/resources.schema.yaml:81]: please elaborate, I don't really understand

Comment (marrip) [reply workflow/schemas/resources.schema.yaml:81]: Sorry, I meant `samtools_sort` and `bam_merge` and so on, shouldn't they be sorted in the schemas like so:
```
default_resources:
   ...
a_rule:
  ...
bam_merge:
  ...
samtools_sort:
  ...
vep:
  ...
```
just an super general simplified example of course.

**Comment by marrip** [comment]
> I have implemented a general sort rule. Please have a look. I had a hard time getting the {file} wildcard to work if we have format {file}.bam (circle dependencies), but with {file}.bam_unsorted it worked.
Makes sense as they are both `.bam`. I guess changing the suffix is a good solution here, right?

File: workflow/schemas/resources.schema.yaml (hunk 5 above)
Comment (Smeds) [reply workflow/schemas/resources.schema.yaml:81]: good suggestion!

**Pull request merged**